wasm-chain-integration/src/v1/trie/api.rs

```diff
@@ -0,0 +1,308 @@
+
+pub type Value = Vec<u8>;
+
+/// The persistent contract state.
+/// Clone on this structure is designed to be cheap, and it is a shallow copy.
+#[derive(Debug, Clone)]
+pub enum PersistentState {
+    Empty,
+    Root(CachedRef<Hashed<Node>>),
+}
+
+impl From<CachedRef<Hashed<Node>>> for PersistentState {
+    fn from(root: CachedRef<Hashed<Node>>) -> Self { Self::Root(root) }
+}
+
+/// Load the persistent state. This only loads the root of the tree. In order to
+/// cache the entire tree into memory use [PersistentTree::cache] afterwards.
+impl Loadable for PersistentState {
+    fn load<S: std::io::Read, F: BackingStoreLoad>(
+        loader: &mut F,
+        source: &mut S,
+    ) -> LoadResult<Self> {
+        match source.read_u8()? {
+            0 => Ok(Self::Empty),
+            1 => Ok(Self::Root(CachedRef::load(loader, source)?)),
+            tag => Err(LoadError::IncorrectTag {
+                tag,
+            }),
+        }
+    }
+}
+
+impl PersistentState {
+    /// Store the tree into the backing store, and modify the tree so that it
+    /// records where parts of it are stored. The root of the tree is written
+    /// into the provided buffer.
+    pub fn store_update_buf<S: BackingStoreStore, W: std::io::Write>(
+        &mut self,
+        backing_store: &mut S,
+        buf: &mut W,
+    ) -> StoreResult<()> {
+        match self {
+            PersistentState::Empty => {
+                buf.write_u8(0)?;
+                Ok(())
+            }
+            PersistentState::Root(node) => {
+                buf.write_u8(1)?;
+                node.store_update_buf(backing_store, buf)
+            }
+        }
+    }
+
+    /// Like [Self::store_update_buf], but returns a freshly allocated buffer
+    /// for the root instead of using the provided one.
+    pub fn store_update<S: BackingStoreStore>(
+        &mut self,
+        backing_store: &mut S,
+    ) -> StoreResult<Reference> {
+        let mut top = Vec::new();
+        self.store_update_buf(backing_store, &mut top)?;
+        backing_store.store_raw(&top)
+    }
+
+    /// Serialize the tree into the provided buffer. Note that this is very
+    /// different from [Self::store_update]. Whereas that stores the part of
+    /// the tree that are only in memory into the provided backing store, this
+    /// function loads the entire tree and serializes it, in a custom format,
+    /// into the provided buffer. As a result this is an expensive operation
+    /// that should only be used when incrementally storing the tree is not
+    /// an option.
+    pub fn serialize(
+        &self,
+        loader: &mut impl BackingStoreLoad,
+        out: &mut impl std::io::Write,
+    ) -> anyhow::Result<()> {
+        match self {
+            PersistentState::Empty => out.write_u8(0)?,
+            PersistentState::Root(ht) => {
+                out.write_u8(1)?;
+                ht.use_value_(loader, |loader, node| node.serialize(loader, out))?
```

Comment:
I agree, I just did not know what the better name is. This pattern that I'm using here is not very rust like. It is very javascripty or Haskelly with supplying a callback to do stuff.

Instead of renaming, I have changed these `use_value` methods to `.get` methods that return something like `Cow` (but not quite that due to needless constraints of that type).